### ISSUE TITLE
[BUG] Severe performance degradation of all commands with versions >=2.24.4 (incl. 2.25.0)

### ISSUE THREAD

**Issue by nocive** [docker/compose#11437]
### Description

Starting with version 2.24.4 I observe a noticeable and severe performance degradation with all docker compose commands.

We have a pretty sizable project with over 100 containers, in case that helps paint a better picture.

### Steps To Reproduce

A big project with all containers stopped (assuming that even matters).

For versions <= 2.24.3
```sh
$ time docker compose ps
NAME      IMAGE     COMMAND   SERVICE   CREATED   STATUS    PORTS


real	0m0.423s
user	0m0.478s
sys	0m0.134s
```

With version >= 2.24.4:
```sh
$ time docker compose ps
NAME      IMAGE     COMMAND   SERVICE   CREATED   STATUS    PORTS


real	0m8.815s
user	0m13.045s
sys	0m0.358s
```

The delay (~8s) is observed before any output is produced and regardless of which command is ran.
I cannot reproduce the slowness with a simple test project of a couple of containers.

Happy to provide more information that can help hunt this down.

### Compose Version

```Text
Docker Compose version v2.24.4
Docker Compose version v2.24.5
Docker Compose version v2.24.6
Docker Compose version v2.24.7
Docker Compose version v2.25.0
```

### Docker Environment

```Text
Client:
 Version:    24.0.7
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  0.12.0
    Path:     /usr/lib/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.3
    Path:     /home/xxx/.docker/cli-plugins/docker-compose

Server:
 Containers: 2121
  Running: 1
  Paused: 0
  Stopped: 2120
 Images: 910
 Server Version: 24.0.7
 Storage Driver: overlay2
  Backing Filesystem: btrfs
  Supports d_type: true
  Using metacopy: true
  Native Overlay Diff: false
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 64b8a811b07ba6288238eefc14d898ee0b5b99ba.m
 runc version: 
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.1.71-1
 Operating System: xxx
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 31.01GiB
 Name: xxx
 ID: xxx
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: xxx
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by nocive** [comment]
perhaps also a symptom of https://github.com/compose-spec/compose-go/pull/557?

**Comment by nocive** [comment]
same problem persists on 2.24.6 :man_shrugging: 
can someone have a look here? :wave:

**Comment by dave-shawley** [comment]
FWIW I'm seeing the same behavior with v2.24.3-desktop.1 on an Apple M2 laptop

```
Docker Compose version v2.24.3-desktop.1

Client:
 Cloud integration: v1.0.35+desktop.10
 Version:           25.0.2
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        29cf629
 Built:             Thu Feb  1 00:18:45 2024
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.27.1 (136059)
 Engine:
  Version:          25.0.2
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       fce6e0c
  Built:            Thu Feb  1 00:23:21 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

This is reproducible with a simple compose.yaml and any docker compose sub-command. The following `docker compose config` took 8 seconds. The interesting thing is that it seems to be sleeping or doing whatever it is up to **after** the output is printed.

```
$ time docker compose config
name: asdf
services:
  nginx:
    image: nginx:1.24-alpine
    networks:
      default: null
    ports:
      - mode: ingress
        target: 80
        protocol: tcp
networks:
  default:
    name: asdf_default
docker compose config  0.08s user 0.03s system 5% cpu 2.106 total
```

**Comment by nocive** [comment]
ping @glours @ndeloof

**Comment by dave-shawley** [comment]
I went a step further in trying to figure this out because having `docker compose ps` take ~10s is killing me. I found that `~/.docker/cli-plugins/docker-compose` is considerably faster than `~/.docker/bin/docker-compose`. I have a feeling that the docker command-line is finding the one in `~/.docker/bin` 🤨 
```
% time (docker compose config>/dev/null)
0.08s user 0.07s system 6% cpu 2.284 total

% time (~/.docker/bin/docker-compose config>/dev/null)
0.08s user 0.07s system 6% cpu 2.276 total

% time (~/.docker/cli-plugins/docker-compose config>/dev/null)
0.03s user 0.01s system 93% cpu 0.048 total
```

Hopefully this will help narrow down where the problem is.

**Comment by glours** [comment]
Hey @nocive @dave-shawley 
Can both of you share the result of `docker compose version`? I wanted to be sure you don't have a old version of Compose taking precedence to the latest one installed because `docker info` tell you the version of the detected Compose plugin but there is not guarantee this is the Compose binary that will be used.
To let you know, I tried to reproduce on my Mac M1 with Desktop in both admin/non-admin mode and with an Unbuntu 23.10 VM without any success

**Comment by nocive** [comment]
I can confirm the correct version is being used:
```
Docker Compose version v2.24.6
```

**Comment by nocive** [comment]
I see a lot of 
```
rt_sigreturn({mask=[]})                 = 1
--- SIGURG {si_signo=SIGURG, si_code=SI_TKILL, si_pid=628483, si_uid=1000} ---
rt_sigreturn({mask=[]})                 = 1
--- SIGURG {si_signo=SIGURG, si_code=SI_TKILL, si_pid=628483, si_uid=1000} ---
rt_sigreturn({mask=[]})                 = 1
--- SIGURG {si_signo=SIGURG, si_code=SI_TKILL, si_pid=628483, si_uid=1000} ---
rt_sigreturn({mask=[]})                 = 1
--- SIGURG {si_signo=SIGURG, si_code=SI_TKILL, si_pid=628483, si_uid=1000} ---
rt_sigreturn({mask=[]})                 = 1
--- SIGURG {si_signo=SIGURG, si_code=SI_TKILL, si_pid=628483, si_uid=1000} ---
rt_sigreturn({mask=[]})                 = 1
--- SIGURG {si_signo=SIGURG, si_code=SI_TKILL, si_pid=628483, si_uid=1000} ---
rt_sigreturn({mask=[]})                 = 1
--- SIGURG {si_signo=SIGURG, si_code=SI_TKILL, si_pid=628483, si_uid=1000} ---
rt_sigreturn({mask=[]})                 = 1
--- SIGURG {si_signo=SIGURG, si_code=SI_TKILL, si_pid=628483, si_uid=1000} ---
rt_sigreturn({mask=[]})                 = 1
--- SIGURG {si_signo=SIGURG, si_code=SI_TKILL, si_pid=628483, si_uid=1000} ---
rt_sigreturn({mask=[]})                 = 1
```

... where a lot of time seems to be spent, and which I don't see in versions <=2.24.3.

**Comment by dave-shawley** [comment]
I reverted back to my installed version since I was seeing the same behavior in v2.24.6
```
% docker compose version
Docker Compose version v2.24.3-desktop.1

% ~/.docker/cli-plugins/docker-compose version
Docker Compose version v2.24.3-desktop.1

% ~/.docker/bin/docker-compose version
Docker Compose version v2.24.3-desktop.1
```

I upgraded to Docker Desktop v4.27.2 this morning which bumped compose to 2.24.5 and the behavior is unchanged.
```
% docker compose version
Docker Compose version v2.24.5-desktop.1

% ~/.docker/bin/docker-compose version
Docker Compose version v2.24.5-desktop.1
```

**Comment by glours** [comment]
@nocive you're on linux? No Docker Desktop for Linux installed? 
Can you try [this Compose binary](https://github.com/docker/compose/actions/runs/7994145230/artifacts/1264075516) and let me know if you see a difference

**Comment by nocive** [comment]
@glours yes, I'm on arch running docker & docker-buildx pkgs + docker compose installed from the releases of this repo :penguin: 

I gave the binary a run and... it became even slower than before :)

Now I see a consistent delay of ~12s instead of ~8s :crying_cat_face:

**Comment by glours** [comment]
@nocive strange 🤔 
Can you try to use a sample from the awesome-compose or a very simple compose project with only few services, and re-run the `docker compose ps` from the sample directory, I just want to check if the performance issues aren't coming from the new version of `compose-go`

**Comment by nocive** [comment]
Possible related/duplicate: https://github.com/docker/compose/issues/11305

**Comment by luciangabor** [comment]
Does running commands with `docker --tls=false compose ...` change anything?
Is there a difference between `docker compose ps` and `docker ps`?

**Comment by nocive** [comment]
```
$ time (source .envrc; docker ps &>/dev/null)

real	0m0.058s
user	0m0.017s
sys	0m0.016s

$ time (source .envrc; docker compose ps &>/dev/null)

real	0m9.797s
user	0m14.592s
sys	0m0.327s

$ time (source .envrc; docker --tls=false compose ps &>/dev/null)

real	0m9.856s
user	0m14.742s
sys	0m0.289s
```

**Comment by nocive** [comment]
When skipping docker compose completely the issue never occurs, which I assume is what you were trying to demonstrate?

```
$ time (source .envrc; docker run --rm -v /var/run/docker.sock:/var/run/docker.sock docker:cli env -i /usr/local/bin/docker version; docker run --rm -v /var/run/docker.sock:/var/run/docker.sock docker:cli env -i /usr/local/bin/docker ps &>/dev/null)

Client:
 Version:           25.0.4
 API version:       1.44
 Go version:        go1.21.8
 Git commit:        1a576c5
 Built:             Wed Mar  6 16:32:02 2024
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          25.0.3
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.22.0
  Git commit:       f417435e5f
  Built:            Thu Feb 29 20:50:13 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.7.13
  GitCommit:        7c3aca7a610df76212171d200ca3811ff6096eb8.m
 runc:
  Version:          1.1.12
  GitCommit:        
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

real	0m1.558s
user	0m0.025s
sys	0m0.037s
```

**Comment by fho** [comment]
The issue has been introduced by the change: https://github.com/compose-spec/compose-go/pull/546

It can be reproduced with [this compose.yaml file](https://gist.githubusercontent.com/fho/3a961e31ad7179f7955107ac8fbb2af6/raw/5332b1803ab634d246b44b48b0b58a83ac06a3d3/compose.yaml).

Running `docker compose config`  on my machine takes with
- **Docker Compose version v2.25.0**: ~6.5s
- **Docker Compose version v2.24.3**: 0,033s
- custom build of v2.25.0 without the https://github.com/compose-spec/compose-go/pull/546: 0,14s.
(much longer than with v2.24.3 but seems acceptable.) 


I created the custom 2.25.0 build without https://github.com/compose-spec/compose-go/pull/546 in the following way:

- Checkout compose.git v2.25.0
- run: `go mod vendor`
- Apply the following change:
   ```diff
   diff --git a/vendor/github.com/compose-spec/compose-go/v2/loader/validate.go b/vendor/github.com/compose-spec/compose-go/v2/loader/validate.go
   index c5133f963..aa1e67192 100644
   --- a/vendor/github.com/compose-spec/compose-go/v2/loader/validate.go
   +++ b/vendor/github.com/compose-spec/compose-go/v2/loader/validate.go
   @@ -76,12 +76,6 @@ func checkConsistency(project *types.Project) error {
                                   return fmt.Errorf("service %q depends on undefined service %q: %w", s.Name, dependedService, errdefs.ErrInvalid)
                           }
                   }
   -               // Check there isn't a cycle in depends_on declarations
   -               if err := graph.InDependencyOrder(context.Background(), project, func(ctx context.Context, s string, config types.ServiceConfig) error {
   -                       return nil
   -               }); err != nil {
   -                       return err
   -               }
   
                   if strings.HasPrefix(s.NetworkMode, types.ServicePrefix) {
                           serviceName := s.NetworkMode[len(types.ServicePrefix):]
   ```
- build a new compose binary: `go build -o compose cmd/main.go`

**Comment by ndeloof** [comment]
I can confirm removing this check has a mass impact on performance. There's something wrong with `InDependencyOrder` implementation, I'll look into it

**Comment by nocive** [comment]
Any chance to get this into master soonish? :pray: :innocent:

**Comment by nocive** [comment]
Fixed in 2.26.0

**Comment by glours** [comment]
Yeah 🎉
Thanks @ndeloof for taking care of this when I was away 🙏🤗
